include/deal.II/lac/linear_operator.h

```diff
@@ -0,0 +1,1175 @@
+                               "Domain>::Tvmult_add called"));
+    };
+
+    reinit_range_vector = [](Range &, bool)
+    {
+      Assert(false, ExcMessage("Uninitialized LinearOperator<Range, "
+                               "Domain>::reinit_range_vector method called"));
+    };
+
+    reinit_domain_vector = [](Domain &, bool)
+    {
+      Assert(false, ExcMessage("Uninitialized LinearOperator<Range, "
+                               "Domain>::reinit_domain_vector method called"));
+    };
+  }
+
+  /**
+   * Create a LinearOperator object from an object @p op for which the
+   * conversion function <code>linop</code> is defined.
+   */
+  template<typename Op>
+  LinearOperator (const Op &op)
+  {
+    *this = linop<Range, Domain, Op>(op);
+  }
+
+  /**
+   * Copy assignment operator for an object @p op for which the conversion
+   * function <code>linop</code> is defined.
+   */
+  template <typename Op>
+  LinearOperator<Range, Domain> &operator=(const Op &op)
+  {
+    return *this = linop<Range, Domain, Op>(op);
+  }
+
+  /**
+   * Application of the LinearOperator object to a vector u of the @tref
+   * Domain space giving a vector v of the @tref Range space.
+   */
+  std::function<const void(Range &v, const Domain &u)> vmult;
+
+  /**
+   * Application of the LinearOperator object to a vector u of the @tref
+   * Domain space. The result is added to the vector v.
+   */
+  std::function<const void(Range &v, const Domain &u)> vmult_add;
+
+  /**
+   * Application of the transpose LinearOperator object to a vector u of
+   * the @tref Range space giving a vector v of the @tref Domain
+   * space.
+   */
+  std::function<const void(Domain &v, const Range &u)> Tvmult;
+
+  /**
+   * Application of the transpose LinearOperator object to a vector @p u of
+   * the @tref Range space.The result is added to the vector @p v.
+   */
+  std::function<const void(Domain &v, const Range &u)> Tvmult_add;
+
+  /**
+   * Initializes a vector v of the Range space to be directly usable
+   * as the destination parameter in an application of vmult. Similar to
+   * the reinit functions of the vector classes, the boolean determines
+   * whether a fast initalization is done, i.e., if it is set to false the
+   * content of the vector is set to 0.
+   */
+  std::function<void(Range &v, bool fast)> reinit_range_vector;
+
+  /**
+   * Initializes a vector of the Domain space to be directly usable as the
+   * source parameter in an application of vmult. Similar to the reinit
+   * functions of the vector classes, the boolean determines whether a fast
+   * initalization is done, i.e., if it is set to false the content of the
+   * vector is set to 0.
+   */
+  std::function<void(Domain &v, bool fast)> reinit_domain_vector;
+
+  /**
+   * A memory object for vectors of @tref Range space used for intermediate
+   * storage during computations in @ref vmult, etc.
+   */
+  mutable GrowingVectorMemory<Range> range_vector_memory;
```

Comment:
I'm also not very happy with those vector memory objects being public.

But making them private is in principle possible, but forces the following changes:
 - add friend declarations for every <code>operator?</code> and <code>inverse_linop</code>
 - encapsulate the vector memory objects in another struct as mutable members to work around the lack of "init capture" facilities in C++11 (comes with C++14)
 - additionally init capture by reference/value the memory object in every lambda expression

Thus, I decided to just keep them as public members in the struct. It at least fits the concept: <code>LinearOperator</code> is a small struct around 4 <code>std::function</code> objects and 2 <code>GrowingVectorMemory</code> objects that can be used/modified at will. (Even partially discarded - individual <code>std::function</code> objects remain valid).